udata/core/dataset/models.py

```diff
@@ -271,7 +313,77 @@ def next_update(self):
             delta = timedelta(weeks=52 * 3)
         elif self.frequency == 'quinquennial':
             delta = timedelta(weeks=52 * 5)
-        return self.last_update + delta
+        if delta is None:
+            return
+        else:
+            return self.last_update + delta
+
+    @cached_property
+    def quality(self):
+        """Return a dict filled with metrics related to the inner
+
+        quality of the dataset:
+
+            * number of tags
+            * description length
+            * and so on
+        """
+        result = {}
+        if self.next_update:
+            result['frequency'] = self.frequency
+            result['update_in'] = -(self.next_update - datetime.now()).days
+        if self.tags:
+            result['tags_count'] = len(self.tags)
+        if self.description:
+            result['description_length'] = len(self.description)
+        if self.resources:
+            result['has_resources'] = True
+            result['has_only_closed_formats'] = all(
+                resource.closed_format for resource in self.resources)
+            result['has_unavailable_resources'] = not all(
+                self.check_availability())
+        discussions = DatasetDiscussion.objects(subject=self)
```

Comment:
This is the faulty line: when I perform an harvest preview, dataset are not persisted.
Si, if you try to query with self, it's not possible.
You should check that `self.id` is set before performing the query.